v2/tools/generator/internal/conversions/README.md

```diff
@@ -0,0 +1,568 @@
+            astmodel.DebugDescription(
+                sourceEndpoint.Type(), 
+                conversionContext.Types()),
+            astmodel.DebugDescription(
+                destinationEndpoint.Type(), 
+                conversionContext.Types()))
+    }
+    if conversion == nil {
+        return nil, nil
+    }
+```
+
+With all our prerequisites satisfied, we return our conversion function.
+
+``` go 
+    return func(
+        reader dst.Expr, 
+        writer func(dst.Expr) []dst.Stmt, 
+        knownLocals *astmodel.KnownLocalsSet, 
+        generationContext *astmodel.CodeGenerationContext) []dst.Stmt {
+``` 
+
+As we did above in `assignFromOptional`, we want to keep the quality of our generated code high by avoiding a local variable if we don't need one.
+
+``` go
+        var cacheOriginal dst.Stmt
+        var actualReader dst.Expr
+
+        switch reader.(type) {
+        case *dst.Ident, *dst.SelectorExpr:
+            // reading a local variable or a field
+            cacheOriginal = nil
+            actualReader = reader
+        default:
+            // Something else, so we cache the original
+            local := knownLocals.CreateSingularLocal(
+                sourceEndpoint.Name(), 
+                "", 
+                "Cache")
+            cacheOriginal = astbuilder.ShortDeclaration(
+                local, 
+                reader)
+            actualReader = dst.NewIdent(local)
+        }
+```
+
+* `cacheOriginal` is an optional statement that reads our value and stashes it in a local variable.
+* `actualReader` is an expression for use later in the function. It reads from our local variable (if used), or directly from our source (if not).
+
+Our array may be nil, so we need to check and avoid dereferencing it if so.
+
+``` go
+        checkForNil := astbuilder.AreNotEqual(actualReader, astbuilder.Nil())
+```
+
+Using `knownLocals`, we create three obviously related identifiers to use for the array conversion. Each uses the name of our source as a base, with a different suffix.
+
+We use different suffixes here to those used in `assignMapFromMap` in order to keep the code easy to read if the two are nested.
+
+``` go
+        branchLocals := knownLocals.Clone()
+        tempId := branchLocals.
+            CreateSingularLocal(sourceEndpoint.Name(), "List")
+
+        loopLocals := branchLocals.Clone()
+        itemId := loopLocals.
+            CreateSingularLocal(sourceEndpoint.Name(), "Item")
+        indexId := loopLocals.
+            CreateSingularLocal(sourceEndpoint.Name(), "Index")
+```
+
+* `List` is the final list we're creating
+* `Item` is a single item from the source list
+* `Index` is the integer index of the item we're converting
+
+Clones of `knownLocals` are used for each nested scope (one inside our **if** statement, see below; the other inside our loop) so that definitions don't leak out.
+
+Our declaration initializes a new list of the correct final size.
+
+``` go
```

Comment:
FWIW I think the bits that are the hardest to understand are the recursive calls and how the endpoints are crafted/work with that process. (Including `knownLocals`, etc). The actual generation of the for loop code once you have the inner conversion isn't all that complicated, so I feel like if you wanted to shorten this document you could remove some of the specificity here and focus on the content you have up above.

It's pretty minor though and you've also already written this so it's probably more work to rejigger it at this point.

Reply:
Another good idea - also noted for a followup PR.